### PR TITLE
Easier preconditions syntax and support for memory preconditions

### LINKED ISSUE #5: Memory preconditions
Analog to the current approach for records, the teacher should be able to specify arbitrary values to be loaded in memory:

```yml
examples:
- preconditions:
    memory:
      A023: '28E5'
      0002: '0003'
```

### LINKED ISSUE #18: Remove memory and records keys from preconditions
To be consistent with the postconditions, the 'memory' and 'records' keys should disappear, since we can distinguish them from their name (the 'R' prefix for records, the special records well-known names, etc).

A migration of the existing exercises will be necessary, following this example:

**Before**
```yml
examples:
- preconditions:
    memory:
      A023: '0002'
   records:
      R2: '0003'
      R3: '0004'
```

**After**
```yml
examples:
- preconditions:
    A023: '0002'
    R2: '0003'
    R3: '0004'
```

The JSON output sent to QSim should be the same as now:

```json
{
  "id": 0,
  "special_records": {
    "PC": "0000",
    "SP": "FFEF",
    "IR": "0000"
  },
  "flags": {
    "N": 0,
    "Z": 0,
    "V": 0,
    "C": 0
  },
  "records": {
    "R0": "0000",
    "R1": "0000",
    "R2": "0003",
    "R3": "0000",
    "R4": "0004",
    "R5": "0000",
    "R6": "0000",
    "R7": "0000"
  },
  "memory": {
    "A023": "0002"
  }
}
```

### PR DESCRIPTION
**Pull request by Charlyzzz**
Closes #18 
Closes #5

### CODE REVIEW

**Comment by faloi** [comment]
The `closes #...` should be in the description, it doesn't work if you put in the title. Besides, a title with just issues numbers is a little rough, you need to read the code to discover its purpose.

File: lib/test_hook.rb [hunk 1]
```diff
@@ -41,8 +43,40 @@ def post_process_file(file, result, status)
...
-    examples.each_with_index.map { |example, index| defaults.merge(example).merge(id: index) }
+    examples.each_with_index.map do |example, index|
+      example[:preconditions] = classify(example.fetch(:preconditions, {}))
+      example.merge(id: index)
+    end
+  end
+
+  def classify(fields)
+    classified_fields = {}
+    fields.map do |key, value|
```
Comment (faloi) [inline lib/test_hook.rb]: Since you don't care about the result of this, you should use `each` instead of `map`.

File: lib/test_hook.rb [hunk 2]
```diff
@@ -41,8 +43,40 @@ def post_process_file(file, result, status)
...
+      field = key.to_s
+      classified_fields.deep_merge!(records: { key => value }) if record?(field)
```
Comment (faloi) [inline lib/test_hook.rb]: You can avoid this repeated logic by creating a function that takes the key and returns its category. Then, you can do something like:

```ruby
kind = classify field 
unless kind == :unknown
  classified_fields[kind].merge! { key => value }
end
```

**Review by faloi — CHANGES_REQUESTED** [review]
Once this is ready and before it's deployed, you should prepare a migration to adapt existing exercises to the new format. 

There is a rake task on bibliotheca that will help you with the migration, see: https://github.com/mumuki/mumuki-bibliotheca-api/blob/master/migrations/migrate_convert_exercise_layout_editors.rb.

File: lib/test_hook.rb (hunk 2 above)
Comment (Charlyzzz) [reply lib/test_hook.rb]: If the kind doesn´t exists yet it is nil, so it won´t look so clean as your example. I'll see what I can do

**Review by faloi — APPROVED** [review]

**Pull request merged**